File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp
class Solution {
public:
    int solveRec(int row, int col, vector<vector<int>>& matrix){
        
        if(row<0 || row>=matrix.size() || col<0 || col>=matrix[0].size()){
            return 1e8;
        }
        if(row == 0){
            return matrix[row][col];
        }
        
        int up = matrix[row][col] + solveRec(row-1, col, matrix);
        int upLeft = matrix[row][col] + solveRec(row-1, col-1, matrix);
        int upRight = matrix[row][col] + solveRec(row-1, col+1, matrix);
        
        return min({up, upLeft, upRight});
    }
    
    int solveMem(int row, int col, vector<vector<int>>& matrix, vector<vector<int>>& dp){
        
        if(row<0 || row>=matrix.size() || col<0 || col>=matrix[0].size()){
            return 1e8;
        }
        if(row == 0){
            return matrix[row][col];
        }
        if(dp[row][col] != -1) return dp[row][col];
        
        int up = matrix[row][col] + solveMem(row-1, col, matrix, dp);
        int upLeft = matrix[row][col] + solveMem(row-1, col-1, matrix, dp);
        int upRight = matrix[row][col] + solveMem(row-1, col+1, matrix, dp);
        
        return dp[row][col] = min({up, upLeft, upRight});
    }
    
    int minFallingPathSum(vector<vector<int>>& matrix) {
        
        int res = INT_MAX;
        int n = matrix.size();
        
        vector<vector<int>> dp(n, vector<int>(n, -1));
        for(int i=0; i<n; i++){
            res = min(res, solveMem(n-1, i, matrix, dp));
        }
        
        return res;
    }
};

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    public int solve(int i, int j, int[][] matrix, int[][] dp){
        
        if(j<0 || j>=matrix.length)
            return 1000005;
        
        if(i == 0){
            return matrix[i][j];
        }
        
        if(dp[i][j] != -1)
            return dp[i][j];
        
        int downLeft = matrix[i][j] + solve(i-1, j-1, matrix, dp);
        int downRight = matrix[i][j] + solve(i-1, j+1, matrix, dp);
        int down = matrix[i][j] + solve(i-1, j, matrix, dp);
        
        return dp[i][j] = Math.min(downLeft, Math.min(down, downRight));
        
    }
    
    public int minFallingPathSum(int[][] matrix) {
        
        int n = matrix.length;
        
        int[][] dp = new int[n][n];
        for(int j=0; j<n; j++)
            dp[0][j] = matrix[0][j];
        
        for(int i=1; i<n; i++){
            
            for(int j=0; j<n; j++){
                
                int downLeft = 1000005, down=1000005, downRight=1000005;
                if(j>=1) downLeft = matrix[i][j] + dp[i-1][j-1];
                if(j+1<n) downRight = matrix[i][j] + dp[i-1][j+1];
                down = matrix[i][j] + dp[i-1][j];

                dp[i][j] = Math.min(downLeft, Math.min(down, downRight));
            }
        }
        
        int ans = 1000005;
        for(int j=0; j<n; j++){
            ans = Math.min(ans, dp[n-1][j]);
        }
        
        return ans;
    }
}